File: app/src/main/java/com/letter/socketassistant/connection/TcpClientConnection.java
package com.letter.socketassistant.connection;

import android.util.Log;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.Socket;

public class TcpClientConnection extends Thread {

    private final int BUFFER_SIZE = 2048;
    private final int PACKET_TIMEOUT = 100;

    private String serverIp;
    private int port;

    private Socket socket;
    private BufferedReader reader;
    private OutputStream outputStream;
    private String line;
    private char[] buffer = new char[BUFFER_SIZE];
    private int index;
    private long time;
    private ConnectionReceivedListener connectionReceivedListener;

    private final String TAG = "TCP CLIENT";

    public TcpClientConnection (String serverIp, int port) {
        this.serverIp = serverIp;
        this.port = port;
    }

    @Override
    public void run() {
        Log.d(TAG, "run: connected start");
        try {
            socket = new Socket(serverIp, port);
            reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
            outputStream = socket.getOutputStream();
            Log.d(TAG, "TcpClientConnection: Connected");

            while (true) {
                if (reader.ready()) {
                    time = System.currentTimeMillis();
                    while (System.currentTimeMillis() - time < PACKET_TIMEOUT
                            && index < BUFFER_SIZE
                            && reader.ready()) {
                        buffer[index++] = (char)reader.read();
                    }
                    if (connectionReceivedListener != null
                            && index > 0) {
                        line = String.valueOf(buffer, 0, index);
                        Log.d(TAG, "run: packet length: " + String.valueOf(index) + ", data: " + line);
                        connectionReceivedListener.onReceivedListener(line);
                    }
                    index = 0;
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            try {
                if (reader != null) {
                    reader.close();
                }
                if (outputStream != null) {
                    outputStream.close();
                }
                if (socket != null) {
                    socket.close();
                }
                Log.d(TAG, "TcpClientConnection: Connect failed");
            } catch (Exception e1) {
                e1.printStackTrace();
            }
        }
    }

    public void setConnectionReceivedListener(ConnectionReceivedListener connectionReceivedListener) {
        this.connectionReceivedListener = connectionReceivedListener;
    }

    public void write(String string) {
        try {
            outputStream.write(string.getBytes("utf-8"));
            outputStream.flush();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void disconnect() {
        try {
            if (reader != null) {
                reader.close();
            }
            if (outputStream != null) {
                outputStream.close();
            }
            if (socket != null) {
                socket.close();
            }
        } catch (Exception e1) {
            e1.printStackTrace();
        }
    }
}

File: app/src/main/java/com/letter/socketassistant/connection/TcpClientConnection.kt
package com.letter.socketassistant.connection

import android.util.Log
import com.letter.socketassistant.utils.tryRead
import java.net.Socket

/**
 * Tcp 客户端连接
 * @property socket Socket socket
 * @property maxPacketLen Int 单个数据包最大长度
 * @property packetTimeOut Long 数据包接收超时
 * @constructor 构造一个连接
 *
 * @author Letter(nevermindzzt@gmail.com)
 * @since 1.0.0
 */
class TcpClientConnection constructor(private var socket: Socket? = null,
                                      private val maxPacketLen: Int = 1024,
                                      private val packetTimeOut: Long = 100)
    : AbstractConnection() {

    companion object {
        private const val TAG = "TcpClientConnection"
    }

    constructor(remoteIp: String?,
                remotePort: Int,
                maxPacketLen: Int = 1024,
                packetTimeOut: Long = 100)
            : this(maxPacketLen=maxPacketLen, packetTimeOut=packetTimeOut) {
        try {
            socket = Socket(remoteIp, remotePort)
            name = "tcp client: ${socket?.inetAddress?.hostAddress}:${socket?.port}"
        } catch (e: Exception) {
            onDisConnectedListener?.invoke(this)
        }
    }

    init {
        name = "tcp client: ${socket?.inetAddress?.hostAddress}:${socket?.port}"
    }

    override fun send(connection: AbstractConnection, bytes: ByteArray?) {
        if (bytes == null) {
            return
        }
        val outputStream = socket?.getOutputStream() !!
        try {
            outputStream.write(bytes)
            outputStream.flush()
        } catch (e : Exception) {
            Log.e(TAG, "", e)
        }
    }

    override fun run() {
        Log.d(TAG, "tcp client run: ${socket?.port}")
        val inputStream = socket?.getInputStream()
        val data = ByteArray(maxPacketLen)
        while (!isInterrupted) {
            try {
                while (inputStream?.available() == 0) {
                    sleep(10)
                }
                val length = inputStream?.tryRead(data, maxPacketLen, packetTimeOut)
                if (length != null) {
                    onReceivedListener?.invoke(this, data.sliceArray(IntRange(0, length - 1)))
                }
            } catch (e: Exception) {
                Log.w(TAG, "", e)
                break
            }
        }
        onDisConnectedListener?.invoke(this)
    }

    override fun disconnect() {
        super.disconnect()
        socket?.getOutputStream()?.close()
        socket?.close()
        Log.d(TAG, "connection disconnect")
    }
}